### ISSUE TITLE
Reduce memory usage in the build

### ISSUE THREAD

**Issue by blink1073** [jupyterlab/jupyterlab#2251]
As reported by @Analect on gitter: https://gist.github.com/Analect/2863bc621471e58d2079d0688a67ea71.

The key is the "process out of memory" error.

Options from looking at open WebPack issues relating to memory:

- Stop using the `ExtractTextPlugin` - we had already planned to do this to enable theming.
- Use  `cheap-source-map`
- Set the memory limit for node `node --max_old_space_size=4096`

https://github.com/webpack/webpack/issues/1914
https://github.com/webpack/webpack/issues/2157

**Comment by blink1073** [comment]
Marking as beta for at least attempting the first two fixes.

**Comment by Analect** [comment]
@blink1073 ... did this get in to `0.22.1`?
I tried again on a 15GB machine ... gets stuck at same spot for circa 10 mins .. and eventually fails.

```
Step 32/42 : RUN jupyter serverextension enable --py jupyterlab --sys-prefix
 ---> Running in de379b226026
Enabling: jupyterlab
- Writing config: /opt/conda/etc/jupyter
    - Validating...
      jupyterlab  OK
 ---> db8b8d78e7d7
Removing intermediate container de379b226026
Step 33/42 : RUN jupyter labextension install jupyterlab-hub
 ---> Running in 1e78fe40b017
```
... sticks at this point for 10 mins ... then resumes and fails with this error.
```
> npm pack jupyterlab-hub
> npm install /opt/conda/share/jupyter/lab/extensions/jupyterlab-hub-0.2.0.tgz
> npm install
> npm run build
b'\n> @jupyterlab/application-top@0.4.1 build /opt/conda/share/jupyter/lab/staging\n> webpack\n\nFATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - process out of memory\n\n<--- Last few GCs --->\n\n  441438 ms: Mark-sweep 1376.8 (1457.6) -> 1379.7 (1457.6) MB, 1674.1 / 0 ms [allocation failure] [GC in old space requested].\n  443115 ms: Mark-sweep 1379.7 (1457.6) -> 1379.6 (1457.6) MB, 1677.0 / 0 ms [allocation failure] [GC in old space requested].\n  444818 ms: Mark-sweep 1379.6 (1457.6) -> 1374.3 (1457.6) MB, 1702.8 / 0 ms [last resort gc].\n  446513 ms: Mark-sweep 1374.3 (1457.6) -> 1376.8 (1457.6) MB, 1695.1 / 0 ms [last resort gc].\n\n\n<--- JS stacktrace --->\n\n==== JS stack trace =========================================\n\n    2: arguments adaptor frame: 3->1\nSecurity context: 0x2c8e0eeb4629 <JS Object>\n    3: InnerArrayForEach(aka InnerArrayForEach) [native array.js:~924] [pc=0xae3324ab0db] (this=0x2c8e0ee041b9 <undefined>,bk=0x39ce6d06d091 <JS Function (SharedFunctionInfo 0x8b1f2421369)>,bl=0x266a67b4ccf9 <a ReplaceSource with map 0x2ea78b9ef5c1>,o=0x3bc9eb309371 <JS Array[8]>,v=8)\n    4: forEach [native array.js:~943] [pc=0xae3325e8841] (this=0x3bc9eb309...\n\nAborted (core dumped)\n\nnpm ERR! Linux 4.4.0-43-generic\nnpm ERR! argv "/opt/conda/bin/node" "/opt/conda/bin/npm" "run" "build"\nnpm ERR! node v4.4.0\nnpm ERR! npm  v2.14.20\nnpm ERR! code ELIFECYCLE\nnpm ERR! @jupyterlab/application-top@0.4.1 build: `webpack`\nnpm ERR! Exit status 134\nnpm ERR! \nnpm ERR! Failed at the @jupyterlab/application-top@0.4.1 build script \'webpack\'.\nnpm ERR! This is most likely a problem with the @jupyterlab/application-top package,\nnpm ERR! not with npm itself.\nnpm ERR! Tell the author that this fails on your system:\nnpm ERR!     webpack\nnpm ERR! You can get information on how to open an issue for this project with:\nnpm ERR!     npm bugs @jupyterlab/application-top\nnpm ERR! Or if that isn\'t available, you can get their info via:\nnpm ERR! \nnpm ERR!     npm owner ls @jupyterlab/application-top\nnpm ERR! There is likely additional logging output above.\n\nnpm ERR! Please include the following file with any support request:\nnpm ERR!     /opt/conda/share/jupyter/lab/staging/npm-debug.log\n'
```

**Comment by blink1073** [comment]
No, this hasn't been released yet.  Although if it is happening on a 15GB machine, I'm not sure what could be going wrong, since I measured 0.5GB peak RAM before #2257.

**Comment by blink1073** [comment]
Are you able to inspect the version of `node` in the docker image?

**Comment by Analect** [comment]
Sorry ... missed this earlier.
So my Dockerfile includes this step `RUN conda install -c javascript nodejs`, which has worked fine with building jupyterlab up to now.  This uses this package: https://anaconda.org/javascript/nodejs ... and it ended up install v4.4.0.
However, I can't altogether reconcile that semantic-versioning with the v7.10.0 on nodejs.org.

Perhaps I should be using the [conda-forge install](https://github.com/conda-forge/nodejs-feedstock) instead, which appears to be at 6.10.2 ... which at least is more in line with the LTS version from nodejs.org.

**Comment by Analect** [comment]
OK ... it looks like that was the source of my problem ... by switching from `RUN conda install -c javascript nodejs` to `RUN conda install -c conda-forge nodejs`, it ends up installing node 6.10.2 and this appears to fix things!   I've no idea why that conda version of nodejs lags so far behind.  Thanks for facilitating a resolution.

**Comment by blink1073** [comment]
We are using node 5 in our Travis tests, with 4GB of RAM available.  Not sure why node 4 would have a problem.  The official anaconda packages tend to move slower, the node packages are about a year old.  I kicked of a build to try the Travis node version 4 to rule that out: https://travis-ci.org/blink1073/jupyterlab/builds/235822620.

**Comment by blink1073** [comment]
Whelp, that build failed to even run webpack on our core packages.  It looks like our minimum supported node is 5.

**Comment by Analect** [comment]
Hmm.  Well at least I know I'm not going mad now ... not yet, at least!

**Comment by blink1073** [comment]
Fixed by #2257 and #2312.
